opm/simulators/flow/rescoup/ReservoirCouplingMaster.cpp

```diff
@@ -0,0 +1,355 @@
+            std::runtime_error,
+            fmt::format(
+                "Slave index {} out of bounds. Valid range is [0, {})",
+                slave_idx, this->slave_idx_to_master_groups_.size()
+            )
+        );
+    }
+
+    // Direct O(1) vector access for performance-critical usage
+    return this->slave_idx_to_master_groups_[slave_idx];
+}
+
+template <class Scalar>
+std::size_t
+ReservoirCouplingMaster<Scalar>::
+getMasterGroupCanonicalIdx(
+    const std::string &slave_name, const std::string &master_group_name) const
+{
+    // NOTE: The master group name order is the canonical order of the master group names
+    //       for a given slave name. This is the order in which the slave will communicate
+    //       slave group data for its slave groups.
+    return this->master_group_name_order_.at(slave_name).at(master_group_name);
+}
+
+template <class Scalar>
+void
+ReservoirCouplingMaster<Scalar>::
+initTimeStepping()
+{
+    assert(!this->time_stepper_);
+    this->time_stepper_ = std::make_unique<ReservoirCouplingTimeStepper<Scalar>>(*this);
+    assert(!this->report_step_data_);
+    this->report_step_data_ = std::make_unique<ReservoirCouplingMasterReportStep<Scalar>>(*this);
+}
+
+template <class Scalar>
+void
+ReservoirCouplingMaster<Scalar>::
+initStartOfReportStep(int report_step_idx)
+{
+    assert(this->report_step_data_);
+    this->report_step_data_->setReportStepIdx(report_step_idx);
+    this->logger_.info("Initializing start of report step");
+}
+
+template <class Scalar>
+bool
+ReservoirCouplingMaster<Scalar>::
+isMasterGroup(const std::string &group_name) const
+{
+    return this->master_group_slave_names_.find(group_name) !=
+           this->master_group_slave_names_.end();
+}
+
+template <class Scalar>
+void
+ReservoirCouplingMaster<Scalar>::
+maybeActivate(int report_step) {
+    if (!this->activated()) {
+        double start_date = this->schedule_.getStartTime();
+        auto current_time = start_date + this->schedule_.seconds(report_step);
+        if (Seconds::compare_gt_or_eq(current_time, this->activation_date_)) {
+            this->activated_ = true;
+        }
+    }
+}
+
+template <class Scalar>
+double
+ReservoirCouplingMaster<Scalar>::
+maybeChopSubStep(double suggested_timestep, double current_time) const
+{
+    assert(this->time_stepper_);
+    return this->time_stepper_->maybeChopSubStep(suggested_timestep, current_time);
+}
+
+template <class Scalar>
+void
+ReservoirCouplingMaster<Scalar>::
+maybeSpawnSlaveProcesses(int report_step)
+{
+    if (this->numSlavesStarted() > 0) {  // We have already spawned the slave processes
+        return;
+    }
+    const auto& rescoup = this->schedule_[report_step].rescoup();
+    auto slave_count = rescoup.slaveCount();
+    auto master_group_count = rescoup.masterGroupCount();
+    if (slave_count > 0 && master_group_count > 0) {
+        ReservoirCouplingSpawnSlaves<Scalar> spawn_slaves{*this, rescoup};
+        spawn_slaves.spawn();
+    }
```

Comment:
Nitpicks:
The if statement could be hidden inside spawn
Not a big one and there might be a good reason for the current way: We are constructing a temporary object to call one function. Could this be just one free function.